### PR TITLE
[Feature] Function Calling

### PR DESCRIPTION
**Pull request by Tushar-ml**
## Motivation

This PR provides support for function calling #2429 . Currently function tooling implemented for tokenizer apply_chat_template. Streaming not support as well for function tooling.

## Checklist

- [x] Format your code according to the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/references/contributor_guide.md).
- [x] Add unit tests as outlined in the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/references/contributor_guide.md).
- [x] Update documentation as needed, including docstrings or example tutorials.

### CODE REVIEW

File: python/sglang/srt/openai_api/adapter.py [hunk 1]
```diff
@@ -1031,11 +1050,51 @@ def v1_chat_generate_response(request, ret, to_file=False, cache_report=False):
...
+                )
+
         if to_file:
             # to make the choice data json serializable
             choice_data = {
                 "index": 0,
-                "message": {"role": "assistant", "content": ret_item["text"]},
+                "message": {
+                    "role": "assistant",
+                    "content": ret_item["text"],
```
Comment (HaoyuWang4188) [inline python/sglang/srt/openai_api/adapter.py]: Maybe we can set content to be `"content": null` just like the [official OpenAI API Reference](https://platform.openai.com/docs/api-reference/chat/create)?

Comment (HaoyuWang4188) [inline python/sglang/srt/openai_api/adapter.py]:
```suggestion
                    "content": ret_item["text"] if(tool_calls is None) else None,
```

File: python/sglang/srt/openai_api/adapter.py [hunk 2]
```diff
@@ -1047,7 +1106,9 @@ def v1_chat_generate_response(request, ret, to_file=False, cache_report=False):
         else:
             choice_data = ChatCompletionResponseChoice(
                 index=idx,
-                message=ChatMessage(role="assistant", content=ret_item["text"]),
+                message=ChatMessage(
+                    role="assistant", content=ret_item["text"], tool_calls=tool_calls
+                ),
```
Comment (HaoyuWang4188) [inline python/sglang/srt/openai_api/adapter.py]:
```suggestion
                message=ChatMessage(
                    role="assistant",
                    content=ret_item["text"] if (tool_calls is None) else None,
                    tool_calls=tool_calls,
                ),
```

**Comment by HaoyuWang4188** [comment]
Based on the communication in Slack, I have left some of my review comments. 😄
Would @Ying1123 please assign an official reviewer? Thank you!

File: docs/references/function_calling.md [hunk 3]
```diff
@@ -0,0 +1,155 @@
+# Tools Calling
+
+Currently, we added the support for tools calling in the following models:
+  - Llama 3.2 models
+  - Llama 3.1 models
+  - Qwen 2.5 models
+  - InternLM Models
+
+For adding support of more different architectures: \n
+ 1. Add Tool tag present in model in line 1055 `sglang/srt/openai_api/adapter.py`
```
Comment (merrymercy) [inline docs/references/function_calling.md]: the line number will change, so do not mention the exact line number.

File: docs/references/function_calling.md [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+client = OpenAI(api_key='YOUR_API_KEY',base_url='http://0.0.0.0:30000/v1')
+model_name = client.models.list().data[0].id
+response = client.chat.completions.create(
+    model=model_name,
+    messages=messages,
+    temperature=0.8,
+    top_p=0.8,
+    stream=False,
+    tools=tools)
+print(response)
```
Comment (merrymercy) [inline docs/references/function_calling.md]: 1. (Necessary) Add a test case in https://github.com/sgl-project/sglang/blob/main/test/srt/test_openai_server.py. Assert the correctness of the output
2. (Optionally) convert this doc into an interactive notebook, similar to the others https://github.com/sgl-project/sglang/blob/main/docs/start/send_request.ipynb

File: docs/references/function_calling.md [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+
+Currently, we added the support for tools calling in the following models:
+  - Llama 3.2 models
+  - Llama 3.1 models
+  - Qwen 2.5 models
+  - InternLM Models
+
+For adding support of more different architectures: \n
+ 1. Add Tool tag present in model in line 1055 `sglang/srt/openai_api/adapter.py`
+ 2. Add support in parsing function for converting into tool calls `sglang/srt/utils.py`
```
Comment (merrymercy) [inline docs/references/function_calling.md]: mention the function name

File: docs/references/function_calling.md [hunk 6]
```diff
@@ -0,0 +1,155 @@
+# Tools Calling
+
+Currently, we added the support for tools calling in the following models:
+  - Llama 3.2 models
+  - Llama 3.1 models
+  - Qwen 2.5 models
+  - InternLM Models
+
+For adding support of more different architectures: \n
```
Comment (merrymercy) [inline docs/references/function_calling.md]: Reorganize this. Mention the usage first and then mention how to support a new model.

```
# Supported models
# Usage
# How to support a new model
```

File: python/sglang/srt/openai_api/adapter.py [hunk 7]
```diff
@@ -1031,11 +1050,51 @@ def v1_chat_generate_response(request, ret, to_file=False, cache_report=False):
...
+                )
+
         if to_file:
             # to make the choice data json serializable
             choice_data = {
                 "index": 0,
-                "message": {"role": "assistant", "content": ret_item["text"]},
+                "message": {
+                    "role": "assistant",
+                    "content": ret_item["text"] if (tool_calls is None) else None,
```
Comment (merrymercy) [inline python/sglang/srt/openai_api/adapter.py]:
```suggestion
                    "content": ret_item["text"] if tool_calls is None else None,
```

File: python/sglang/srt/openai_api/adapter.py [hunk 8]
```diff
@@ -1047,7 +1106,11 @@ def v1_chat_generate_response(request, ret, to_file=False, cache_report=False):
         else:
             choice_data = ChatCompletionResponseChoice(
                 index=idx,
-                message=ChatMessage(role="assistant", content=ret_item["text"]),
+                message=ChatMessage(
+                    role="assistant",
+                    content=ret_item["text"] if (tool_calls is None) else None,
```
Comment (merrymercy) [inline python/sglang/srt/openai_api/adapter.py]:
```suggestion
                    content=ret_item["text"] if tool_calls is None else None,
```

File: python/sglang/srt/openai_api/adapter.py [hunk 9]
```diff
@@ -1031,11 +1050,51 @@ def v1_chat_generate_response(request, ret, to_file=False, cache_report=False):
...
+
+        if isinstance(request, list):
+            tool_choice = request[idx].tool_choice
+            tools = request[idx].tools
+        else:
+            tool_choice = request.tool_choice
+            tools = request.tools
+
+        if tool_choice != "none" and (
+            "<|plugin|>" in text
```
Comment (merrymercy) [inline python/sglang/srt/openai_api/adapter.py]: 1. Make these constants a list of constant strings
2. Put this list above `parse_tool_response` in `python/sglang/srt/utils.py`
3. Import this list here.

After this, people only need to update one file to support a new model.

**Review by merrymercy — COMMENTED** [review]
This looks good! I left a few comments.

**Comment by Tushar-ml** [comment]
@merrymercy @Ying1123 The above comments have been resolved

File: docs/backend/function_tooling.ipynb [hunk 10]
```diff
@@ -0,0 +1,116 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Function Tooling\n",
```
Comment (merrymercy) [inline docs/backend/function_tooling.ipynb]:
```suggestion
    "# Function Calling\n",
```

Update the title and filename in all places.
https://platform.openai.com/docs/guides/function-calling

File: docs/references/function_calling.md [hunk 11]
```diff
@@ -0,0 +1,61 @@
+# Tools Calling
```
Comment (merrymercy) [inline docs/references/function_calling.md]: Merge `docs/references/function_calling.md` and `docs/backend/function_tooling.ipynb`.
I think we only need the content of this file but in the format of an interactive notebook.

The outline of this doc is very good!

File: docs/references/function_calling.md [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+'''
+
+ChatCompletion(id='d6f620e1767e490d85b5ce45c15151cf', choices=[Choice(finish_reason='stop', index=0, logprobs=None, message=ChatCompletionMessage(content=None, refusal=None, role='assistant', audio=None, function_call=None, tool_calls=[ChatCompletionMessageToolCall(id='0', function=Function(arguments='{"a": "3", "b": "5"}', name='add'), type='function')]), matched_stop=128008)], created=1735411703, model='meta-llama/Llama-3.2-1B-Instruct', object='chat.completion', service_tier=None, system_fingerprint=None, usage=CompletionUsage(completion_tokens=23, prompt_tokens=198, total_tokens=221, completion_tokens_details=None, prompt_tokens_details=None))
+
+'''
+```
+
+## How to support a new model?
+
+For adding support of more different models: \n
```
Comment (merrymercy) [inline docs/references/function_calling.md]: "\n" is not needed.

**Review by merrymercy — CHANGES_REQUESTED** [review]
Some final minor comments! Then we can merge this.

**Comment by Tushar-ml** [comment]
Done @merrymercy

**Pull request merged**